pennylane/transforms/optimization/pattern_matching.py

```diff
@@ -0,0 +1,1567 @@
+# Copyright 2021-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transform finding all maximal matches of a pattern in a quantum circuit and optimizing the circuit by
+substitution."""
+
+import itertools
+import copy
+import pennylane as qml
+from pennylane import apply
+from pennylane.transforms import qfunc_transform
+from pennylane.transforms.commutation_dag import commutation_dag
+from pennylane.wires import Wires
+from pennylane.ops.qubit.attributes import (
+    symmetric_over_all_wires,
+)
+
+
+@qfunc_transform
+def pattern_matching(tape, pattern_tapes):
+    r"""Quantum function transform to optimize a circuit given a list of patterns (templates).
+
+    Args:
+        qfunc (function): A quantum function to be optimized.
+        pattern_tapes(list(.QuantumTape)): List of quantum tapes that implement the identity.
```

Comment:
Thank you for your comment, I don't think it is perfect because it is less known than quantum function. But the patterns are just a succession of operations that implement the identity and I think it is the best candidate for this purpose. I could also use quantum function but those should not have any returns.

The transform will not be integrated in the default pipeline as I see it more with something that you run only once on your circuit and should save the optimized circuit. This algorithm is time consuming and should not be run every time and also highly depends on the chosen pattern.